src/store/actions/updateBalances.ts

```diff
@@ -1,108 +1,251 @@
+      if (chains[acc.chain].evmCompatible) {
+        const nativeAsset = chains[acc.chain].nativeAsset;
+        const client = getters.client({
+          network,
+          walletId,
+          asset: nativeAsset,
+          accountId: a,
+        }) as Client<EvmChainProvider>;
+
+        // add only EVM chains that has multicall support
+        if (client.chain.multicall) {
+          if (!result[acc.chain]) {
+            result[acc.chain] = [];
           }
-        } else {
-          updatedAddresses = [...addresses.map((a) => a.toString())];
+          result[acc.chain].push(acc);
         }
+      }
+    }
+
+    return result;
+  }, {} as EvmAccountsMap);
+};
+
+const updateEVMBalances = async (
+  context: ActionContext,
+  evmAccounts: EvmAccountsMap,
+  network: Network,
+  walletId: WalletId
+) => {
+  const { getters, commit } = rootActionContext(context);
+
+  await Bluebird.map(
+    Object.entries(evmAccounts),
+    async ([chain, accounts]) => {
+      // each account for each chain can be used to get a chainify client
+      const evmAccount = accounts[0];
+      const nativeAsset = chains[chain as ChainId].nativeAsset;
+
+      // common evm client
+      const client = getters.client({
+        network,
+        walletId,
+        asset: nativeAsset,
+        accountId: evmAccount.id,
+      }) as Client<EvmChainProvider>;
+
+      // fetch the address for all accounts
+      const addressesForAllAccounts = await Promise.all(
+        accounts.map((a) => {
+          return getters
+            .client({
+              network,
+              walletId,
+              asset: nativeAsset,
+              accountId: a.id,
+            })
+            .wallet.getUsedAddresses();
+        })
+      );
+
+      // update address for each account
+      accounts.map((acc, index) => {
+        updateAccountAddresses(context, acc, addressesForAllAccounts[index], network, walletId);
+      });
+
+      const multicallCallData = accounts.reduce((result, a, index) => {
+        // evm accounts have only 1 address
+        const user = addressesForAllAccounts[index][0].toString();
 
-        commit.UPDATE_ACCOUNT_ADDRESSES({
+        a.assets.forEach((asset) => {
+          const chainifyAsset = assetsAdapter(asset)[0];
+          const callData = client.chain.multicall.buildBalanceCallData(chainifyAsset, user);
+          result.push(callData);
+        });
+
+        return result;
+      }, [] as EvmTypes.MulticallData[]);
+
+      console.log('multicallCallData', multicallCallData);
+      const result = await client.chain.multicall.multicall<BigNumber[]>(multicallCallData);
+      console.log('result', result);
+
+      accounts.forEach((acc) => {
+        const balances = result.splice(0, acc.assets.length).map((balance, index) => {
+          if (balance) {
+            return balance.toString();
+          } else {
+            console.debug(`Balance not fetched: ${acc.assets[index]}`);
+            return '0';
```

Comment:
IMO setting '0' on failure is not the right outcome. It will mean that if there is a network error, user will suddenly see 0 and might even freak out. Can we do something where failed fetch just does not update the balance for that sset ? This way, previous balance will be preserved.

Reply:
Fixed in https://github.com/liquality/wallet-core/pull/97/commits/f8c7cc7f642eb9b809e4ae0aaf4cd440d8b95d5f